mongo/src/org/immutables/mongo/bson4gson/BsonReader.java

```diff
@@ -72,6 +73,26 @@ public void close() throws IOException {
     this.delegate = checkNotNull(delegate, "delegate");
   }
 
+  @Override
+  public <T> T unwrap(Class<T> type) {
```

Comment:
thinking of if `Wrapper<T>` would be more staightforward and replace  `<T> T unwrap(Class<T> type)` with simple `T unwrap()`. For example,
`class BsonReader extends JsonReader implements Wrapper<org.bson.BsonReader> {`
then if you cast to `BsonReader` in adapter you know that `unwrap()` will get `org.bson.BsonReader`, if you don't cast, but `instanceof Wrapper<?>`, then `unwrap()` will return `Object` and the client will try to instanceof / cast on his own. Also, is it realistic that a delegate itself would be a wrapper in the way we need to care about? I just have a feeling that this piece is too flexible/promising, but actual usages are not much needing this.

Reply:
That's a fair point. 
Wrapper interface was inspired by [JDBC Wrapper](https://docs.oracle.com/javase/7/docs/api/java/sql/Wrapper.html). Perhaps it allows too much (eg. `unwrap(Closeable.class)`)

How about calling it `Delegate<T>` or re-using [ForwardingObject](https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/ForwardingObject.html) ?

Reply:
> Wrapper interface was inspired by JDBC Wrapper. Perhaps it allows too much (eg. unwrap(Closeable.class))

Yep, recognized patter. JDBC do not have a chance provide any type safety via `Wrapper<T>`, so they also decided that it's possible to get different kinds of delegates depending of a passed in class, so they decided to be quite flexible.

> How about calling it Delegate<T> or re-using ForwardingObject (in fact delegate() method is protected and doesn't help much) ?

`Wrapper<T>` or `DelegatesTo<T>` would be just fine, `ForwardingObject` is for their (Guava) needs.

> In general, do you agree with exposing original BsonReader (new way) instead of creating delegate methods (old way) ?

probably yes, don't see much overwhelming benefits on either side, probably less delegation in a new model

Reply:
Addressed your comments